### PR TITLE
refactor: Encapsulates listbox and trigger slot construction

### PR DESCRIPTION
**Pull request by ling1726**
Removes the `useTriggerListboxSlots` hook in favour of separate `useListboxSlot` and `useTriggerSlot` hooks for better encapsulation of each behaviour.

Also stop using `hasFocus` as a method of opening the listbox popover.

### CODE REVIEW

**Comment by fabricteam** [comment]
## 🕵 fluentuiv9 No visual regressions between this PR and main
<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

File: packages/react-components/react-combobox/src/components/Combobox/useCombobox.tsx [hunk 1]
```diff
@@ -156,31 +80,11 @@ export const useCombobox_unstable = (props: ComboboxProps, ref: React.Ref<HTMLIn
...
-  const listboxRef = useMergedRefs(listboxSlot?.ref, comboboxPopupRef);
-
-  if (hideActiveDescendant) {
-    triggerSlot['aria-activedescendant'] = undefined;
-  }
-
-  if (listboxSlot) {
-    listboxSlot.ref = listboxRef;
-  }
+  triggerSlot = useInputTriggerSlot(props, baseState, ref, triggerSlot);
```
Comment (YuanboXue-Amber) [inline packages/react-components/react-combobox/src/components/Combobox/useCombobox.tsx:83]: Why do we have to pass `ref`, can we do:
on L75: `ref: useMergedRefs(props.input?.ref, triggerRef, ref)`,
and in `useInputTriggerSlot` just use triggerSlot.ref?

File: change/@fluentui-react-combobox-05096ff8-98cf-46e5-b371-d525d4e9f4eb.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (fabricteam) [inline change/@fluentui-react-combobox-05096ff8-98cf-46e5-b371-d525d4e9f4eb.json:1]: ## [🕵 fluentuiv9 Open the Visual Regressions report to inspect the affected screenshots]( https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-approval/index.html?organization=uifabric&project=fabricpublic&prId=30010&commitId=f94c098ffe69ac44b6ae14bb1bc74070592a1ebd&env=prod&clientType=FLUENTUI)


  <details><summary><b>Dropdown Converged</b> 34 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Dropdown Converged.Disabled.default.chromium.png|250962| Changed|
|Dropdown Converged.Invalid- outline.default.chromium.png|250688| Changed|
|Dropdown Converged.Appearance- outline (default).default.chromium.png|250816| Changed|
|Dropdown Converged.Disabled option.default.chromium.png|250791| Changed|
|Dropdown Converged.Open with inlinePopup.default.chromium.png|250791| Changed|
|Dropdown Converged.Option with long content.default.chromium.png|250791| Changed|
|Dropdown Converged.Appearance- filled-lighter.default.chromium.png|249995| Changed|
|Dropdown Converged.Open with grouped options.default.chromium.png|250791| Changed|
|Dropdown Converged.Appearance- underline.default.chromium.png|250651| Changed|
|Dropdown Converged.Appearance- filled-darker.default.chromium.png|249683| Changed|
|Dropdown Converged.Invalid- filled-darker.default.chromium.png|249367| Changed|
|Dropdown Converged.Invalid- underline.default.chromium.png|250653| Changed|
|Dropdown Converged.Disabled with value.default.chromium.png|250943| Changed|
|Dropdown Converged.Open.default.chromium.png|250791| Changed|
|Dropdown Converged.Invalid- filled-lighter.default.chromium.png|249657| Changed|
|Dropdown Converged.Size- large.default.chromium.png|251560| Changed|
|Dropdown Converged.With multiselect value.default.chromium.png|250634| Changed|
|Dropdown Converged.Size- small.default.chromium.png|249252| Changed|
|Dropdown Converged.With placeholder.default.chromium.png|250675| Changed|
|Dropdown Converged.With multiselect selection.default.chromium.png|250791| Changed|
|Dropdown Converged.With selection.default.chromium.png|250791| Changed|
|Dropdown Converged.With value.default.chromium.png|250684| Changed|
|Dropdown Converged.Disabled option.active option.chromium.png|0| Removed|
|Dropdown Converged.Disabled option.hover.chromium.png|0| Removed|
|Dropdown Converged.Open with grouped options.active option.chromium.png|0| Removed|
|Dropdown Converged.Open with grouped options.hover.chromium.png|0| Removed|
|Dropdown Converged.Open with inlinePopup.active option.chromium.png|0| Removed|
|Dropdown Converged.Open with inlinePopup.hover.chromium.png|0| Removed|
|Dropdown Converged.Open.active option.chromium.png|0| Removed|
|Dropdown Converged.Open.hover.chromium.png|0| Removed|
|Dropdown Converged.Option with long content.active option.chromium.png|0| Removed|
|Dropdown Converged.Option with long content.hover.chromium.png|0| Removed|
|Dropdown Converged.With selection.active option.chromium.png|0| Removed|
|Dropdown Converged.With selection.hover.chromium.png|0| Removed|

</details>


  <details><summary><b>Field</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Field.Dropdown.default.chromium.png|328828| Changed|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

File: packages/react-components/react-combobox/src/components/Combobox/useCombobox.tsx (hunk 1 above)
Comment (ling1726) [reply packages/react-components/react-combobox/src/components/Combobox/useCombobox.tsx]: Yeah also, the `ref` is already in the slot at that point - there's no point even having a ref argument either

**Review by YuanboXue-Amber — APPROVED** [review]

**Comment by fabricteam** [comment]
## 🕵 fluentuiv9 No visual regressions between this PR and main
<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

File: packages/react-components/react-combobox/src/components/Dropdown/useButtonTriggerSlot.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
+import * as React from 'react';
+import { useTimeout, type ExtractSlotProps, type Slot, mergeCallbacks } from '@fluentui/react-utilities';
```
Comment (layershifter) [inline packages/react-components/react-combobox/src/components/Dropdown/useButtonTriggerSlot.ts:2]: As we don't compile down types, will it cause issues with older TS versions?

File: packages/react-components/react-combobox/src/utils/useListboxSlot.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  );
+
+  const onClick = useEventCallback(
+    mergeCallbacks((event: React.MouseEvent<HTMLDivElement>) => {
+      event.preventDefault();
+      triggerRef.current?.focus();
+    }, listboxSlot?.onClick),
+  );
+
+  // resolve listbox shorthand props
```
Comment (layershifter) [inline packages/react-components/react-combobox/src/utils/useListboxSlot.ts:34]:
```suggestion
```

Nothing happens there

File: packages/react-components/react-combobox/src/components/Dropdown/useDropdown.tsx [hunk 5]
```diff
@@ -36,72 +34,19 @@ export const useDropdown_unstable = (props: DropdownProps, ref: React.Ref<HTMLBu
...
+  let listbox = slot.optional(props.listbox, {
+    renderByDefault: true,
+    defaultProps: { children: props.children },
+    elementType: Listbox,
+  });
+  const listboxRef = useMergedRefs(listbox?.ref, comboboxPopupRef);
+  if (listbox) {
+    listbox.ref = listboxRef;
+  }
+  listbox = useListboxSlot(props, listbox, triggerRef) as typeof listbox;
```
Comment (layershifter) [inline packages/react-components/react-combobox/src/components/Dropdown/useDropdown.tsx:47]: Why not following?

```ts
const listbox = useListboxSlot(props, slot.optional(props.listbox, {
    renderByDefault: true,
    defaultProps: { children: props.children },
    elementType: Listbox,
  }), triggerRef);
const listboxRef = useMergedRefs(listbox?.ref, comboboxPopupRef);

  if (listbox) {
    listbox.ref = listboxRef;
  }
```

Or even (do `slot.optional(props.listbox` in `useListboxSlot()`):

```ts
const listbox = useListboxSlot(props, triggerRef);
const listboxRef = useMergedRefs(listbox?.ref, comboboxPopupRef);

  if (listbox) {
    listbox.ref = listboxRef;
  }
```

Comment (ling1726) [reply packages/react-components/react-combobox/src/components/Dropdown/useDropdown.tsx]: With some help from @bsunderhus, we figured out the interface types for slots

**Pull request merged**